proof/refine/ARM/SchedContext_R.thy

```diff
@@ -185,6 +185,46 @@ lemma scConsumed_update_corres:
   apply (clarsimp simp: sc_relation_def objBits_def objBitsKO_def obj_at_def)
   done
 
+lemma scBadge_update_corres:
+  "\<lbrakk>\<forall>x y. x = y \<longrightarrow> f x = f' y\<rbrakk> \<Longrightarrow>
+    corres dc (sc_at scPtr) (ko_at' sc' scPtr)
+          (update_sched_context scPtr (sc_badge_update f))
+          (setSchedContext scPtr (scBadge_update f' sc'))"
+  apply (clarsimp simp: update_sched_context_def)
+  apply (rule corres_symb_exec_l[rotated 2, OF get_object_sp])
+    apply (find_goal \<open>match conclusion in "\<lbrace>P\<rbrace> f \<exists>\<lbrace>Q\<rbrace>" for P f Q \<Rightarrow> -\<close>)
+    apply (fastforce intro: get_object_exs_valid
+                      simp: obj_at_def)
+   apply wpsimp
+   apply (clarsimp simp: obj_at_def)
+  apply (rename_tac obj)
+  apply (case_tac obj; clarsimp;
+         (solves \<open>fastforce simp: obj_at_def is_sc_obj_def corres_underlying_def\<close>)?)
+  apply (corressimp corres: setSchedContext_no_stack_update_corres
+                              [where f="sc_badge_update f" and f'="scBadge_update f'" ])
+  apply (clarsimp simp: sc_relation_def objBits_def objBitsKO_def obj_at_def)
+  done
+
+lemma scSporadic_update_corres:
+  "\<lbrakk>\<forall>x y. x = y \<longrightarrow> f x = f' y\<rbrakk> \<Longrightarrow>
+    corres dc (sc_at scPtr) (ko_at' sc' scPtr)
+          (update_sched_context scPtr (sc_sporadic_update f))
+          (setSchedContext scPtr (scSporadic_update f' sc'))"
+  apply (clarsimp simp: update_sched_context_def)
```

Comment:
For this one and the one above (`scBadge_update_corres`), and also `scConsumed_update_corres `, we probably don’t need separate rules for each of these. We should be able to have a generic version to cover all these cases, for corresponding between `setSchedContext` and `update_sched_context`, when updating non-reply, non-refills fields.